packages/react-utilities/src/compose/getSlots.ts

```diff
@@ -1,8 +1,88 @@
 import * as React from 'react';
 
-import { getNativeElementProps, omit } from '../utils/index';
-import { GenericDictionary } from './types';
+import { ComponentState, ShorthandRenderFunction, SlotPropsRecord } from './types';
 import { nullRender } from './nullRender';
+import { getNativeElementProps, omit } from '../utils/index';
+
+function getSlot(
+  defaultComponent: React.ElementType | undefined,
+  userComponent: keyof JSX.IntrinsicElements | undefined,
+  alternativeComponent: React.ElementType,
+) {
+  if (defaultComponent === undefined || typeof defaultComponent === 'string') {
+    return userComponent ?? defaultComponent ?? alternativeComponent;
+  }
+  return defaultComponent;
+}
+
+/**
+ * Given the state and an array of slot names, will break out `slots` and `slotProps`
+ * collections.
+ *
+ * The root is derived from a mix of `components` props and `as` prop.
+ *
+ * Slots will render as null if they are rendered as primitives with undefined children.
+ *
+ * The slotProps will always omit the `as` prop within them, and for slots that are string
+ * primitives, the props will be filtered according the the slot type. For example, if the
+ * slot is rendered `as: 'a'`, the props will be filtered for acceptable anchor props.
+ *
+ * @param state - State including slot definitions
+ * @param slotNames - Name of which props are slots
+ * @returns An object containing the `slots` map and `slotProps` map.
+ */
+export function getSlots<SlotProps extends SlotPropsRecord = {}>(
+  state: ComponentState<SlotProps>,
+  slotNames?: Array<keyof SlotProps>,
+) {
+  /**
+   * force typings on state, this should not be added directly in parameters to avoid type inference
+   */
+  const typedState = state as ComponentState<SlotProps>;
+  type Slots = { [K in keyof SlotProps]: React.ElementType<SlotProps[K]> };
+  const slots = {} as Slots;
+  const slotProps = {} as SlotProps;
+
+  if (slotNames) {
+    for (const name of slotNames) {
+      const { as, children, ...rest } = typedState[name];
+
+      slots[name] = getSlot(typedState.components?.[name], as, 'span') as Slots[typeof name];
```

Comment:
why is the `alternativeComponent` for all non-root slots a `span` which is different from the root variant ?

Reply:
Well, that's a good question. I copied from previous implementationm you can check it out in `getSlotCompat`, `div` is used for root and `span` for all the rest of the slots

Reply:
If that's not right we could remove it... that would definitely make my life easier

Reply:
having different kinds of defaults in root vs other slots seems inconsistent to me, you'd literally have to code dig the day you ever wondered why your root slot is a div but other slot is a span

Reply:
@layershifter , commented something similar in a previous conversation. Should I alter this for `div` everywhere?! I'm just following previous implementation here 😅

Reply:
> Should I alter this for div everywhere?!

Yes, let's use `div` by default everywhere. It's written in the spec, so let's adjust the implementation